### ISSUE TITLE
parakeet TDT greedy decode: blank + duration-0 forces frame advance, diverging from NeMo/MLX reference

### ISSUE THREAD

**Issue by pszemraj** [CrispStrobe/CrispASR#88]
hi, coming from [this comment](https://github.com/CrispStrobe/CrispASR/issues/81#issuecomment-4428643538), the first of issues I alluded to: after some more extended usage for PTT dictation via [my daemon-style CLI ](https://github.com/pszemraj/parakit)using your codebase as a backend, I started to notice sometimes a few words (up to a sentence) would be missing from the end of a dictation when I know the hardware/pipeline had enough time to handle it after I let go of PTT hotkey

Had claude investigate:

---

## Summary

The TDT greedy decode loop in `parakeet_tdt_decode()` handles blank predictions differently from the NeMo and parakeet-mlx reference implementations. When the joint head predicts blank with duration 0, CrispASR force-advances the frame pointer by 1. The reference implementations keep the decoder on the same encoder frame and retry the joint step, only force-advancing after `max_symbols` (typically 10) consecutive zero-duration steps.

This _likely_ causes end-of-sequence truncation on long utterances. Phrases or full sentences near the tail of 30-45+ second dictations are silently dropped.

## Reproduction

Feed a 30-45 second continuous dictation (natural speech with pauses, not read text) through `parakeet_transcribe_ex()`. Compare the output against the same audio processed by [parakeet-mlx](https://github.com/senstella/parakeet-mlx) or [NeMo](https://github.com/pszemraj/parakit/blob/da20f698ed5937ace7f3ad34216d1bfa12f39d47/scripts/transcribe_nemo_parakeet.py). The CrispASR output is often missing the last 1-3 seconds of speech. The truncation is not consistent across utterances -- it depends on the model's blank/duration distribution at the tail frames.

## Root cause

`src/parakeet.cpp` line 1090-1093 (commit `acda0622`):

```cpp
if (tok == blank_id) {
    t += std::max(1, dur_skip);   // blank+dur0 -> advance by 1
    break;                         // exit inner loop
}
```

The `max(1, dur_skip)` means blank predictions always advance `t` by at least 1 frame. The inner loop (`n_inner`, line 1023) never runs for blanks because the `break` fires first.

Compare with the reference in parakeet-mlx `parakeet.py` lines 586-611 (commit `ba03a1b6`):

```python
step += self.durations[int(decision)]   # dur=0 -> step stays put

new_symbols += 1
if self.durations[int(decision)] != 0:
    new_symbols = 0
else:
    if self.max_symbols is not None and self.max_symbols <= new_symbols:
        step += 1
        new_symbols = 0
```

Key differences:

- **Blank + dur 0**: Reference stays on the same frame and retries. CrispASR advances by 1 and never retries.
- **Stuck prevention**: Reference counts all zero-duration predictions (blank and non-blank) toward a shared `max_symbols` counter (typically 10 from the NeMo greedy config: `decoding.greedy.max_symbols`). CrispASR's `max_per_step` counter (line 969, value 10) only counts non-blank emissions with dur 0.

The TDT model sometimes emits blank + dur 0 as a "not yet decided" signal, expecting a retry at the same frame to resolve to a real token on the next joint pass. The forced advancement eliminates that retry. Each skipped frame is 80ms. Over the tail of a long sequence where the decoder is less confident, 20-30 prematurely skipped frames produce 1.6-2.4 seconds of lost speech.

## Suggested fix

Replace the blank handling at line 1090-1093 with NeMo-compatible retry logic:

```cpp
if (tok == blank_id) {
    if (dur_skip > 0) {
        t += dur_skip;
        break;
    }
    // dur_skip == 0: stay on this frame, count toward stuck limit
    n_inner++;
    if (n_inner >= max_per_step) {
        t++;
        break;
    }
    continue;
}
```

This keeps the existing `max_per_step = 10` as the stuck limit (matching the typical NeMo `max_symbols` default), but lets blank + dur 0 retry instead of force-advancing.

The non-blank path (lines 1109-1142) does not need changes. The fix is scoped to the blank branch only.

## Environment

- CrispASR `main` at `acda0622`
- Compared against parakeet-mlx `v0.5.1` at `ba03a1b6`
- NeMo reference configs: `fastconformer/hybrid_transducer_ctc/fastconformer_hybrid_tdt_ctc_bpe.yaml`
- Model: `nvidia/parakeet-tdt-0.6b-v3` (Q4_K GGUF)
- Observed on Windows, CPU/OpenBLAS, ~45 second dictation-style recordings

**Comment by CrispStrobe** [comment]
Thanks for the detailed write-up — the diff against `GreedyTDTInfer._greedy_decode` in NeMo and the parakeet-mlx outer-loop snippet made this very easy to triage.

Landed your fix (with one small adjustment) in [`3a9b294c`](https://github.com/CrispStrobe/CrispASR/commit/3a9b294c). The blank handler now matches NeMo:

```cpp
if (tok == blank_id) {
    if (dur_skip > 0) { t += dur_skip; break; }
    n_inner++;
    continue;   // stay on this frame; while-condition + post-loop handle exhaustion
}
```

### One nit on the proposed patch

Your suggested form adds an explicit force-advance inside the blank handler:

```cpp
if (n_inner >= max_per_step) { t++; break; }
```

That double-increments `t` when blank+dur=0 exhausts the budget — the existing post-loop check at line 1145 fires again, jumping the encoder ahead by 2 frames instead of 1. Dropping the in-handler force-advance and letting the while-condition + post-loop check do their job gives the cleanest match to NeMo's `time_idx += 1` after `symbols_added == max_symbols`.

### Whether this fixes the truncation

For pure greedy argmax (the default — `decode_temperature == 0`) the two implementations are observationally equivalent on the decode loop alone. Blank predictions don't mutate predictor state, so on retry `_pred_step` and `joint_step` get identical inputs and produce identical logits → identical blank+dur=0. The NeMo path then spins `max_symbols` times producing the same blank, and the post-loop `if symbols_added == max_symbols: time_idx += 1` advances by exactly 1 frame. CrispASR's old code advanced by 1 frame immediately. **Same final hypothesis, same emitted tokens.**

So: the patch is correct and worth landing for reference parity + sampling correctness + forward-compat, but I'd be surprised if it alone explains end-of-utterance truncation on long greedy dictation. Mental model from your write-up — "the TDT model emits blank+dur=0 as a 'not yet decided' signal expecting a retry to resolve to a real token" — only holds when the joint output can change between retries, which requires either temperature sampling or some other state mutation between iterations. Neither happens on the default greedy path.

Concrete asks if you can:

1. **Re-run your 30-45 s reproducer on this commit and tell me whether the tail truncation actually goes away.** If it does, great — there's likely a subtle non-determinism I haven't traced. If it doesn't, we should look elsewhere.
2. **Share a problematic WAV** (or a short snippet of one) if possible. I'd love to diff `parakeet_emitted_token` positions against parakeet-mlx's hypothesis frame-by-frame on the same input. The encoder side is already diff-tested (`tools/dump_reference.py --backend parakeet` captures `encoder_output`), so we can isolate exactly where divergence starts. The decoder side has no reference dump yet — adding one for the TDT decode trace is on my list now.
3. **Confirm your `decode_temperature` setting in parakit.** If you're passing anything > 0, the old blank+dur=0 force-advance was demonstrably wrong (and this patch fixes it). If you're at 0 and still see truncation, that's a separate bug worth chasing.

Either way — appreciate the careful diagnosis with line numbers and reference URLs. Made the fix straightforward.

Also: I'm going to look at the encoder output length / VAD-padding side of things in parallel. PTT dictation has a specific failure mode where the trailing silence after speech-end can be encoder-trimmed before the decoder sees the last few frames containing real speech (the encoder's `length` arithmetic is `(audio_len // hop) // subsampling` and the floor division can drop up to ~1 frame at boundaries). If your hardware/pipeline reliably "had enough time" after PTT release, the more likely root cause is on the audio framing path, not the decode loop.

I'll ping back here if I find a second bug worth fixing.

**Comment by pszemraj** [comment]
Thanks! Looks fixed to me btw

I'm using your v0.6.6 iirc in the latest version of Parakit that I finished a branch for over the weekend and I believe it's resolved
